File: backend-springboot/src/main/java/br/com/portalproposta/backend/controller/UsersController.java
package br.com.portalproposta.backend.controller;

import br.com.portalproposta.backend.model.CreateUsuarioRequest;
import br.com.portalproposta.backend.model.UpdateUsuarioRequest;
import br.com.portalproposta.backend.model.Usuario;
import br.com.portalproposta.backend.model.UsuarioResponse;
import br.com.portalproposta.backend.service.UsuarioService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.servlet.support.ServletUriComponentsBuilder;

import java.net.URI;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/users")
public class UsersController {

    private final UsuarioService usuarioService;

    @Autowired
    public UsersController(UsuarioService usuarioService) {
        this.usuarioService = usuarioService;
    }

    @GetMapping
    public ResponseEntity<List<UsuarioResponse>> list() {
        List<Usuario> users = usuarioService.list();
        List<UsuarioResponse> response = users.stream().map(Usuario::toResponse).toList();
        return ResponseEntity.ok(response);
    }

    @PostMapping
    public ResponseEntity<UsuarioResponse> create(@RequestBody CreateUsuarioRequest req) {
        Usuario user = Usuario.builder()
                .nome(req.getNome())
                .email(req.getEmail())
                .perfilId(req.getPerfilId())
                .ativo(req.isAtivo())
                .build();

        Usuario created = usuarioService.create(user, req.getSenha());
        URI uri = ServletUriComponentsBuilder.fromCurrentRequest()
                .path("/{id}")
                .buildAndExpand(created.getId())
                .toUri();
        return ResponseEntity.created(uri).body(created.toResponse());
    }

    @GetMapping("/{id}")
    public ResponseEntity<UsuarioResponse> get(@PathVariable String id) {
        Usuario user = usuarioService.getById(id);
        return ResponseEntity.ok(user.toResponse());
    }

    @PutMapping("/{id}")
    public ResponseEntity<UsuarioResponse> update(@PathVariable String id, @RequestBody UpdateUsuarioRequest req) {
        Usuario input = Usuario.builder()
                .nome(req.getNome())
                .email(req.getEmail())
                .perfilId(req.getPerfilId())
                .ativo(req.isAtivo())
                .build();

        Usuario updated = usuarioService.update(id, input, req.getSenha());
        return ResponseEntity.ok(updated.toResponse());
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Map<String, String>> delete(@PathVariable String id) {
        usuarioService.delete(id);
        Map<String, String> response = new HashMap<>();
        response.put("message", "Usuário excluído com sucesso");
        return ResponseEntity.ok(response);
    }
}

File: backend/Controllers/UsersController.cs
using Microsoft.AspNetCore.Mvc;
using PortalProposta.Models;
using PortalProposta.Services;
using System.Text.Json.Serialization;

namespace PortalProposta.Controllers;

[ApiController]
[Route("api/users")]
public class UsersController(UsuarioService usuarioService) : ControllerBase
{
    public record CreateUsuarioRequest(
        [property: JsonPropertyName("nome")] string Nome,
        [property: JsonPropertyName("email")] string Email,
        [property: JsonPropertyName("senha")] string Senha,
        [property: JsonPropertyName("perfil_id")] string PerfilId,
        [property: JsonPropertyName("ativo")] bool Ativo);

    public record UpdateUsuarioRequest(
        [property: JsonPropertyName("nome")] string Nome,
        [property: JsonPropertyName("email")] string Email,
        [property: JsonPropertyName("senha")] string? Senha,
        [property: JsonPropertyName("perfil_id")] string PerfilId,
        [property: JsonPropertyName("ativo")] bool Ativo);

    [HttpGet]
    public async Task<ActionResult<List<UsuarioResponse>>> List()
    {
        var users = await usuarioService.ListAsync();
        return Ok(users.Select(u => u.ToResponse()).ToList());
    }

    [HttpPost]
    public async Task<ActionResult<UsuarioResponse>> Create(CreateUsuarioRequest req)
    {
        var user = new Usuario
        {
            Nome = req.Nome,
            Email = req.Email,
            PerfilId = req.PerfilId,
            Ativo = req.Ativo
        };
        var created = await usuarioService.CreateAsync(user, req.Senha);
        return CreatedAtAction(nameof(Get), new { id = created.Id }, created.ToResponse());
    }

    [HttpGet("{id}")]
    public async Task<ActionResult<UsuarioResponse>> Get(string id) =>
        Ok((await usuarioService.GetByIdAsync(id)).ToResponse());

    [HttpPut("{id}")]
    public async Task<ActionResult<UsuarioResponse>> Update(string id, UpdateUsuarioRequest req)
    {
        var input = new Usuario
        {
            Nome = req.Nome,
            Email = req.Email,
            PerfilId = req.PerfilId,
            Ativo = req.Ativo
        };
        return Ok((await usuarioService.UpdateAsync(id, input, req.Senha)).ToResponse());
    }

    [HttpDelete("{id}")]
    public async Task<IActionResult> Delete(string id)
    {
        await usuarioService.DeleteAsync(id);
        return Ok(new { message = "Usuário excluído com sucesso" });
    }
}
